types/src/transaction/block_epilogue.rs

```diff
@@ -0,0 +1,39 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_crypto::HashValue;
+#[cfg(any(test, feature = "fuzzing"))]
+use proptest_derive::Arbitrary;
+use serde::{Deserialize, Serialize};
+use std::fmt::Debug;
+
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[cfg_attr(any(test, feature = "fuzzing"), derive(Arbitrary))]
+pub enum BlockEpiloguePayload {
+    BlockId(HashValue),
+    WithBlockEndInfo {
+        block_id: HashValue,
+        block_end_info: BlockEndInfo,
+    },
+}
+
+impl BlockEpiloguePayload {
+    pub fn try_as_block_end_info(&self) -> Option<&BlockEndInfo> {
+        match self {
+            BlockEpiloguePayload::BlockId(_) => None,
+            BlockEpiloguePayload::WithBlockEndInfo { block_end_info, .. } => Some(block_end_info),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct BlockEndInfo {
+    /// Whether block gas limit was reached
+    pub block_gas_limit_reached: bool,
+    /// Whether block output limit was reached
+    pub block_output_limit_reached: bool,
+
+    pub block_effective_block_gas: u64,
+
+    pub block_approx_output_size: u64,
```

Comment:
shall we also add transaction_size, and/or other detailed breakdowns, e.g. # of items/bytes added/deleted/updated.

Reply:
@msmouse

Reply:
Without a concrete use case, I'd refrain from adding onchain data. Does the state sync really need these info and these info certified on chain? I'd let @JoshLind speak for himself.

Reply:
+1. I don't think state sync needs it. It might make certain things easier, but without concrete plans/optimizations, it feels a little premature to me. 😄